### ISSUE TITLE
Unexpected timeouts when running Capybara test suites

### ISSUE THREAD

**Issue by geoffharcourt** [heroku/rack-timeout#55]
I'm running a Rails 4.1.6 app on Heroku, developing locally on OSX Yosemite, and I found an issue that might be a regression:

Upon updating my app to use v0.1.0 from v0.0.4, my test suite started hanging on some of my Capybara feature tests. If I roll back to using v0.0.4, my test suite takes under 45 seconds to run. If I re-update to v0.1.0, my test suite never seems to finish (I cut it off after 11 minutes). The slow/delayed tests are all Capybara features. If I mix up the order it doesn't seem to make a difference, but the issue only affects tests that include loading some content by Ajax.

The issue seems to have no effect on my application in staging (running on Heroku), it only manifests itself during tests.

I'm using Capybara 2.4.4 and Capybara-webkit 1.3.0 with RSpec 3.1.7.

**Comment by vasilakisfil** [comment]
Hi, I also have an issue with rack-timeout. Specifically, if I include it in my application, both in tests and development env cause  "Attempt to unlock a mutex which is locked by another thread" errors. Here is an example: http://pastie.org/private/dosnsedmobnfnpmhtpszaa

I use puma, rails 4.2 and ruby 2.2.

**Comment by seanknox** [comment]
Hi, another soul here with the same issue: feature/integration tests with Capybara hang, no apparent issue on Heroku apps directly. Rails 4.2.0 app using turbolinks and light backbone.js, rack-timeout v0.1.0, capybara/selenium, postgres. To reiterate: very light client-side code. It's almost entirely pure Rails.

- Feature specs with selenium/capybara often hang—still trying to track down all the variables when this started in our codebase but it has worsened. Difficult because this is nondeterministic, but using rspec seeds, I can reproduce hangs pretty reliably within a run or two.
- In development, we also occasionally have seen `ThreadError: Attempt to unlock a mutex which is locked by another thread; occurred at /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rack-1.6.0/lib/rack/lock.rb:22:in `unlock' `. This definitely happened with WEBrick, but I don't think it has happened with thin or (now) puma. Not 100%, though.
- Non-feature specs (non-Capybara) never hang—controllers, routing specs, models, etc.—thousands of test runs confirm this.
- After removing `rack-timeout` I could **not** reproduce a hang after 50+ test suite runs. :white_check_mark: 
- After changing Capybara's server to Thin (it's WEBrick by default) I could **not** reproduce a hang after 50+ test suite runs. :white_check_mark: 
- After downgrading `rack-timeout` to 0.0.4, I could **not** reproduce a hang after 50+ test suite runs. :white_check_mark: 

Again, our test environment is quite hangtastic at the moment that it's odd to not have a hang within a few test runs.

Here's a snippet from the test log during a hang. In this example `Rack::Timeout.timeout` is cranked up to 20s:

```
source=rack-timeout id=4e1dd1deb63d9c9aa55df596af9a6c13 timeout=20000ms service=18ms state=completed
source=rack-timeout id=a32c18826a396926b77f81309082963b timeout=20000ms state=ready
Started GET "/assets/application.css" for 127.0.0.1 at 2015-02-22 00:36:33 -0800
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms state=ready
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=1ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=1005ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=2005ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=3004ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=4002ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=5000ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=6002ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=7005ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=8000ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=9002ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=10000ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=11000ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=12000ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=13002ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=14004ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=15005ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=16005ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=17003ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=18005ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=19003ms state=active
source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=20005ms state=timed_out
source=rack-timeout id=a32c18826a396926b77f81309082963b timeout=20000ms service=2ms state=completed
```

Notice request id `a32c18826a396926b77f81309082963b` attempts to download `/assets/application.css`. Immediately following is another request that will time out: `a5a1716dd69cf02e18e299ac5a7878f2`. It eventually does, and the preceding request for application.css finally completes. Presumably, the timed out request is blocked IO. After a request times out via `rack-timeout`, Capybara never recovers and all subsequent tests fail due to timeouts. 

I'm loathe to remove rack-timeout from our tests entirely as I'd like to mimic our production environment as close as possible in CI.

**Comment by seanknox** [comment]
Update: I believe I've correlated the start of our timeout woes. The events, in order, were:

- Upgraded rack-timeout gem from 0.0.4 to 0.2.0 
- First feature/capybara/Selenium spec introduced (this was a new project)—first observed hang in CI within several days of this. I've jumped back to this commit on my own machine have been able to reproduce a hang. It's much harder and takes many more runs as there is only a single feature spec.
- Increasing frequency of hangs as new feature specs are introduced. In our current codebase with about 20 or so feature specs, a hang can be reproduced easily, within a few test runs, as mentioned in my earlier comment.

I've been diving into this issue for the past week and initially thought it was an ActiveRecord/Database issue as the AR connection would be lost just before the "hung" test, despite AR itself reporting that a connection was active. I have some more information/thoughts on this and can share—hopefully this is enough to get eyeballs on possible culprits.

**Comment by seanknox** [comment]
@vasilakisfil are you using puma as a test rails server with feature specs?

**Comment by kch** [comment]
@seanknox, thanks for digging deeper into this.

> Rails 4.2.0 app using turbolinks and sprinkling of backbone.js, rack-timeout v0.1.0, capybara/selenium, postgres.

Would be great to start stripping down those components. We need to get to a miminal test case here.

If you're willing to put in the time, I'd appreciate if you can make a clone of your app and start ripping things out.

> - Feature specs with selenium/capybara often hang—still trying to track down all the variables when this started in our codebase but it has worsened. Difficult because this is nondeterministic, but using rspec seeds, I can reproduce hangs pretty reliably within a run or two.

> - In development, we also occasionally have seen `ThreadError: Attempt to unlock a mutex which is locked by another thread; occurred at /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rack-1.6.0/lib/rack/lock.rb:22:in 'unlock' `. This definitely happened with WEBrick, but I don't think it has happened with thin or (now) puma. Not 100%, though.

I 've spun this off into another issue: #65. It may or may not be related to this one.

As for WEBrick, I've said it before supporting it is the lowest thing in my list of priorities. No one should dare running it in production and I don't see any advantages to running it locally either. So it's a case of "Doctor it hursts when I do this", "Well, don't do it."

> - Non-feature specs (non-Capybara) do _not_ hang—controllers, routing specs, models, etc.—thousands of test runs confirm this.
> - After removing `rack-timeout` I could **not** reproduce a hang after 50+ test suite runs. 

Good, this is another datapoint to confirm this issue only maifests when Capybara is around, and it's some kind of interaction between Capybara and rack-timeout.

> - After changing Capybara's server to Thin (it's WEBrick by default) I could **not** reproduce a hang after 50+ test suite runs.

Cool. What about puma, threaded and unthreaded?

Everyone else seeing this issue, would be great if you can try the same and let me know if it's a WEBrick thing. I would not be surprised at all (have I mentioned disdain for WEBrick?)


> ```
> …
> source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=1005ms state=active
> source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=2005ms state=active
> source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=3004ms state=active
> source=rack-timeout id=a5a1716dd69cf02e18e299ac5a7878f2 timeout=20000ms service=4002ms state=active
> source=rack-timeout id=a32c18826a396926b77f81309082963b timeout=20000ms service=2ms state=completed
> …
> ```
> 
> … Presumably, the timed out request is blocked IO. 

I originally thought so too, but when rack-timeout is IO-blocked, you don't get the intermediate active state log lines every second because that thread get blocked too. So I don't think it's the case here. And the mutex error kinda signals a potential threading issue.

> After a request times out via `rack-timeout`, Capybara never recovers and all subsequent tests fail due to timeouts.

This is the kinda shit I've seen WEBrick do. Not to bang on it too much without proof, but I think it's a component to keep an eye on.


> Here's what HTTPScoop shows:

I don't see anything in the screenshot that pops out at me, but sometimes the interest stuff is what you see in a live run. Did you notice anything interesting? There may be nothing here after all.

**Comment by kch** [comment]
@seanknox

> … have been able to reproduce a hang. It's much harder and takes many more runs as there is only a single feature spec.
> Increasing frequency of hangs as new feature specs are introduced …

Can you on the same run, just run the same spec multiple times instead of running many different ones?

> I've been diving into this issue for the past week and initially thought it was an ActiveRecord/Database …

I somehow doubt it's db/AR-related. but…

> I have some more information/thoughts on this and can share—hopefully this is enough to get eyeballs on possible culprits.

…please do share and let's see what we can learn.

**Comment by mockdeep** [comment]
> As for WEBrick, I've said it before supporting it is the lowest thing in my list of priorities. No one should dare running it in production and I don't see any advantages to running it locally either. So it's a case of "Doctor it hursts when I do this", "Well, don't do it."

@kch this is a case where nothing "hurt" before upgrading `rack-timeout` so we didn't do that, and it sounds like that was the solution for some of the other people on this thread as well. For us it was simply easier when we were using Unicorn in production to use WEBrick in development, and none of the other servers showed any advantages there. I can understand if you're not particularly motivated to dig into this, but you are harping on a bit about WEBrick without any actual evidence.

At this point we've switched to puma, though, in production, so we should be able to upgrade now...

**Comment by seanknox** [comment]
From what I've read on this issue and my team's experiences, I think Capybara's rack server + rack-timeout 0.2.0 as the main players in this issue.  When people in this issue mention that use a server other than WEBrick, their integration tests with Capybara are probably still using WEBrick, and I'd hazard a guess that their plain `rails s` server is using WEBrick too. `Procfile` settings != `rails s` != Capybara's server. Hope that helps someone.

Overnight I ran my full test suite 1,000 times in an order known to reproduce hangs reliably. The only difference in configuration was using Thin as Capybara's server:

```ruby
Capybara.server do |app, port|
  require 'rack/handler/thin'
  Rack::Handler::Thin.run(app, Port: port)
end
```
I had no hangs or other issues; everything was :peach:y-keen.

@kch agreed that a threading problem is involved, especially since Capybara tests are the only ones that launch a server in a different thread in the same rspec process, causing database shenanigans many of us are familiar with.

> Can you on the same run, just run the same spec multiple times instead of running many different ones?

It's been very hard to reproduce the hangs with a single spec; this is primarily why our team found ourselves in the middle of the hangswamp without realizing it. As more integration specs were added (more Capybara use), hang frequency went up. I can reproduce hangs with a single integration spec when run with the other specs (controllers, models, etc.) but it's still difficult to reproduce.

> Cool. What about puma, threaded and unthreaded?

Using Puma's rack server + Capybara with default thread settings (min: 0, max: 16) I was able to yield a hang with the mutex error. Backtrace:

```ruby
   Failure/Error: Unable to find matching line from backtrace
     ThreadError:
       Attempt to unlock a mutex which is locked by another thread
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rack-1.6.0/lib/rack/lock.rb:22:in `unlock'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rack-1.6.0/lib/rack/lock.rb:22:in `ensure in call'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rack-1.6.0/lib/rack/lock.rb:23:in `call'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/actionpack-4.2.0/lib/action_dispatch/middleware/static.rb:113:in `call'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rack-1.6.0/lib/rack/sendfile.rb:113:in `call'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rack-timeout-0.2.0/lib/rack/timeout.rb:108:in `call'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/railties-4.2.0/lib/rails/engine.rb:518:in `call'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/railties-4.2.0/lib/rails/application.rb:164:in `call'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rack-1.6.0/lib/rack/urlmap.rb:66:in `block in call'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rack-1.6.0/lib/rack/urlmap.rb:50:in `each'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rack-1.6.0/lib/rack/urlmap.rb:50:in `call'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/server.rb:19:in `call'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/puma-2.11.0/lib/puma/server.rb:507:in `handle_request'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/puma-2.11.0/lib/puma/server.rb:375:in `process_client'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/puma-2.11.0/lib/puma/server.rb:262:in `block in run'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/puma-2.11.0/lib/puma/thread_pool.rb:104:in `call'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/puma-2.11.0/lib/puma/thread_pool.rb:104:in `block in spawn_thread'
```

When I started Puma with min:max threads of 0:1, I couldn't reproduce a hang. I'm running on MRI and am not familiar with Puma's threading model, so I'm not sure if these are actual system threads or some virtual threading that Puma does—would love if someone could share/point to info on how that works.

> I don't see anything in the screenshot that pops out at me, but sometimes the interest stuff is what you see in a live run. Did you notice anything interesting? There may be nothing here after all.

The hangs we've noticed always correspond, time-wise, with the browser requesting an asset and never receiving a response. Might be not be important.

**Comment by seanknox** [comment]
Update: occasionally Puma barfs on integration specs that use `Timecop`. Here's one where we're testing password expiration works with devise:

```ruby
Error in reactor loop escaped: Invalid argument (Errno::EINVAL)
/opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/puma-2.11.0/lib/puma/reactor.rb:28:in `select'
/opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/puma-2.11.0/lib/puma/reactor.rb:28:in `run_internal'
/opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/puma-2.11.0/lib/puma/reactor.rb:126:in `block in run_in_thread'
```

The tests keep moving along actually, and rspec reports they pass. If I catch error and cancel the tests manually, full backtrace:

```ruby
Failures:

  1) Password expires user is forced to change his/her password
     Failure/Error: log_out
     EOFError:
       end of file reached
     # /opt/rubies/2.2.0/lib/ruby/2.2.0/net/protocol.rb:153:in `read_nonblock'
     # /opt/rubies/2.2.0/lib/ruby/2.2.0/net/protocol.rb:153:in `rbuf_fill'
     # /opt/rubies/2.2.0/lib/ruby/2.2.0/net/protocol.rb:134:in `readuntil'
     # /opt/rubies/2.2.0/lib/ruby/2.2.0/net/protocol.rb:144:in `readline'
     # /opt/rubies/2.2.0/lib/ruby/2.2.0/net/http/response.rb:39:in `read_status_line'
     # /opt/rubies/2.2.0/lib/ruby/2.2.0/net/http/response.rb:28:in `read_new'
     # /opt/rubies/2.2.0/lib/ruby/2.2.0/net/http.rb:1413:in `block in transport_request'
     # /opt/rubies/2.2.0/lib/ruby/2.2.0/net/http.rb:1410:in `catch'
     # /opt/rubies/2.2.0/lib/ruby/2.2.0/net/http.rb:1410:in `transport_request'
     # /opt/rubies/2.2.0/lib/ruby/2.2.0/net/http.rb:1383:in `request'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/webmock-1.20.4/lib/webmock/http_lib_adapters/net_http.rb:97:in `block in request'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/webmock-1.20.4/lib/webmock/http_lib_adapters/net_http.rb:110:in `call'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/webmock-1.20.4/lib/webmock/http_lib_adapters/net_http.rb:110:in `block in request'
     # /opt/rubies/2.2.0/lib/ruby/2.2.0/net/http.rb:853:in `start'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/webmock-1.20.4/lib/webmock/http_lib_adapters/net_http.rb:109:in `request'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/selenium-webdriver-2.44.0/lib/selenium/webdriver/remote/http/default.rb:83:in `response_for'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/selenium-webdriver-2.44.0/lib/selenium/webdriver/remote/http/default.rb:39:in `request'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/selenium-webdriver-2.44.0/lib/selenium/webdriver/remote/http/common.rb:40:in `call'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/selenium-webdriver-2.44.0/lib/selenium/webdriver/remote/bridge.rb:640:in `raw_execute'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/selenium-webdriver-2.44.0/lib/selenium/webdriver/remote/bridge.rb:618:in `execute'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/selenium-webdriver-2.44.0/lib/selenium/webdriver/remote/bridge.rb:375:in `clickElement'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/selenium-webdriver-2.44.0/lib/selenium/webdriver/common/element.rb:54:in `click'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/selenium/node.rb:73:in `click'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/node/element.rb:133:in `block in click'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/node/base.rb:84:in `synchronize'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/node/element.rb:133:in `click'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/node/actions.rb:27:in `click_link'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/session.rb:676:in `block (2 levels) in <class:Session>'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/dsl.rb:51:in `block (2 levels) in <module:DSL>'
     # ./spec/support/feature_helpers/authentication_helpers.rb:15:in `log_out'
     # ./spec/features/password_expiration_spec.rb:24:in `block (2 levels) in <top (required)>'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/example.rb:177:in `instance_exec'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/example.rb:177:in `block in run'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/example.rb:385:in `block in with_around_and_singleton_context_hooks'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/example.rb:343:in `block in with_around_example_hooks'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/hooks.rb:474:in `block in run'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/hooks.rb:614:in `block in run_around_example_hooks_for'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/example.rb:247:in `call'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/example.rb:247:in `call'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-rails-3.2.0/lib/rspec/rails/adapters.rb:111:in `block (2 levels) in <module:MinitestLifecycleAdapter>'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/example.rb:333:in `instance_exec'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/example.rb:333:in `instance_exec'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/hooks.rb:387:in `execute_with'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/ho/opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/puma-2.11.0/lib/puma/reactor.rb:28:in `select'oks.rb:616:in `block (2 levels) in run_around_example_hooks_for'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/example.rb:247:in `call'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/example.rb:247:in `call'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/hooks.rb:617:in `run_around_example_hooks_for'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/hooks.rb:474:in `run'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/example.rb:343:in `with_around_example_hooks'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/example.rb:385:in `with_around_and_singleton_context_hooks'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/example.rb:174:in `run'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/example_group.rb:548:in `block in run_examples'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/example_group.rb:544:in `map'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/example_group.rb:544:in `run_examples'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/example_group.rb:512:in `run'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/runner.rb:110:in `block (3 levels) in run_specs'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/runner.rb:110:in `map'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/runner.rb:110:in `block (2 levels) in run_specs'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/configuration.rb:1526:in `with_suite_hooks'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/runner.rb:109:in `block in run_specs'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/reporter.rb:62:in `report'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/runner.rb:108:in `run_specs'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/runner.rb:86:in `run'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/runner.rb:70:in `run'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/lib/rspec/core/runner.rb:38:in `invoke'
     # /opt/rubies/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.0/exe/rspec:4:in `<top (required)>'
     # bin/rspec:16:in `load'
     # bin/rspec:16:in `<main>'
```

**Comment by vasilakisfil** [comment]
@seanknox yes

**Comment by seanknox** [comment]
@kch Changing Capybara's rack server to thin seemed to fix locally but still yielded some hangs in our CI (Semaphore). In the meantime we've disabled using `rack-timeout` in `Rails.env.test` which while not ideal, at least unblocks us while the source problem is determined.

**Comment by vasilakisfil** [comment]
Yeah I also have disabled it from dev and test envs, seems the only solution.

**Comment by sambecker** [comment]
I've had the same issues w/out Capybara. I switched from Unicorn to Puma w/ Rack timeouts in order to follow Heroku best practices and immediately had hanging issues on simple page reloads. And these are detrimental hangs that once corrupted a Rails project and always require a full system reboot. Force quitting Puma keeps port 3000 blocked. Also, as has been pointed out, these are silent crashes. Never any console output that would provide any clues. I followed the advice here and moved rack timeout to `:production` only. If it helps I'm running Yosemite 10.10.2, Rails 4.2 and Ruby 2.2.0.

**Comment by kch** [comment]
@sambecker Puma with multiple threads, yes?

When you say you're only using it in production, does that mean the problems you had locally were present if using the rails application in development mode locally, or just when running tests?

**Comment by sambecker** [comment]
@kch Yep. Standard Puma config (see below).

And yes, the problems occurred while running in development mode locally. I was trying to follow Heroku best practices (outlined here, https://devcenter.heroku.com/articles/deploying-rails-applications-with-the-puma-web-server) for having the production environment match the development environment as much as possible.

```
workers Integer(ENV['WEB_CONCURRENCY'] || 2)
threads_count = Integer(ENV['MAX_THREADS'] || 5)
threads threads_count, threads_count

preload_app!

rackup      DefaultRackup
port        ENV['PORT']     || 3000
environment ENV['RACK_ENV'] || 'development'

on_worker_boot do
  # Worker specific setup for Rails 4.1+
  # See: https://devcenter.heroku.com/articles/deploying-rails-applications-with-the-puma-web-server#on-worker-boot
  ActiveRecord::Base.establish_connection
end
```

**Comment by dkubb** [comment]
As mentioned in https://github.com/heroku/rack-timeout/issues/65#issuecomment-86161454 I think I was able to resolve this issue with a few fixes to `Rack::Lock`.

There are details in that comment if anyone wants to help test it out in their environment.

**Comment by etipton** [comment]
#dkubb4prez

**Comment by kch** [comment]
Would be great to here from everyone here if @dkubb's rack fix resolves your issues. More importantly, if it doesn't.

**Comment by seanknox** [comment]
Running test suite with `rack` 1.6.0 + @dkubb's fixes, and rack-timeout enabled with a 5s timeout, has been solid and without issue for several hours. Looking very promising.

**Comment by etipton** [comment]
Hmm, I didn't have any issues using @dkubb's fix but unfortunately I can't recreate my original issue anyway :-/ Sorry I can't truly help test this.

**Comment by jekuno** [comment]
@dkubb I tested your fix using
``gem 'rack', '~> 1.6.0', git: 'https://github.com/dkubb/rack.git', branch: 'fix/1-6-stable/rack-lock-mutex-usage'``
Unfortunately this doesn't fix the problem for me.

For now I worked around this issue by using rack-timeout only in staging and production.

**Comment by jpaulomotta** [comment]
I'm having the exact same issue here.
 No cabybara. Rails 4.2, Ruby 2.2.5,  and rack-timeout 0.2.0

I'm not running integration tests. Most of the time the problem appears when I'm manually testing in development mode. after a few page reloads the whole thing freezes and this pops in my server logs. I always have to restart the server.

Haven't seen the issue in production yet.

**Comment by westoque** [comment]
This recently just worked for me.

Basically just put `rack-timeout` gem in development and production environments fixed it.

However, I still don't understand how it did so.

**Comment by yaneeka** [comment]
@jpaulomotta I'm seeing similar issues. Were you able to pinpoint/workaround your issue since your comment?

**Comment by jpaulomotta** [comment]
@yaneeka, I just uninstalled the gem and I'm not using it in production /
development anymore. I hope you can find a workaround for your issue.

**Comment by westoque** [comment]
@jpaulomotta @yaneeka I actually solved this awhile ago by making sure that rack-timeout middleware is inserted **after** `Rack::Lock`

**Comment by kch** [comment]
Did we ever get that rack lock patch into rack?

(just leaving this here so I'll actually investigate later)

**Comment by yaneeka** [comment]
@westoque Not sure that I understood your answer, perhaps because my usage is different than yours (or my experience is less). Can you explain in what context you are referring to when you say inserted?

**Comment by westoque** [comment]
@yaneeka When you are using `rack-timeout` with Rails. `rack-timeout` automatically inserts itself in the ***0*** position of your rails middleware stack since it is a rack middleware.

Now, the issue that I've found (at least in our case) was that it was causing timeouts because of thread locking ([related issue here](https://github.com/heroku/rack-timeout/issues/65#issuecomment-86161454))

Further investigation leads me to `Rack::Lock`. It turns out `Rack::Lock` is a middleware to make sure that we are ***threadsafe***. So after that, the logical decision was to insert `Rack::Timeout` after `Rack::Lock`.

And this has seemed to fix our problem.

**Comment by kch** [comment]
@westoque how exactly are you doing that? Because I was going to write some documentation on that but I can't get it to work. I'm going the obvious way and using `config.middleware.insert_after Rack::Lock, Rack::Timeout`, but I get ` No such middleware to insert after: Rack::Lock (RuntimeError)`.

`Rack::Lock` is definitely there when I run `rake middleware`. This is a fresh Rails 4.2 test app.

**Comment by kch** [comment]
Ugh Rails. Seems Rails 4 doesn't use `Rack::Lock` anymore. But then why does it show in `rake middleware`. @schneems WHYYY

**Comment by kch** [comment]
Ok, it has to do with `allow_concurrency`. Still, `rake middleware` is stupid.

**Comment by yaneeka** [comment]
Ok, so comment from @westoque wouldn't apply to a Rails 4 app then and my problem is likely elsewhere?

**Comment by kch** [comment]
@yaneeka Rails 4 will still use `Rack::Lock` depending on your choice of server and the `allow_concurrency` setting.

**Comment by kch** [comment]
@yaneeka just pushed some changes to master that change the way rack-timeout is added to rails, so it should be after Rack::Lock, but work with Rails 4 too:

Try sourcing rack-timeout from the [rails-middleware-order](https://github.com/heroku/rack-timeout/tree/rails-middleware-order) tag.

**Comment by kch** [comment]
With #85, I consider this no longer an issue.

**Comment by kch** [comment]
Watch #78 for the next beta.

**Comment by westoque** [comment]
@kch Thanks for the fix. :+1:

**Comment by yaneeka** [comment]
Thanks, I'll give it a go.

**Comment by yaneeka** [comment]
@kch, looks like I'm still running into my problem, so the issue must be elsewhere. Bummer.

**Comment by kch** [comment]
@yaneeka can you open a new issue with details about your particular problem?

**Comment by yaneeka** [comment]
@kch, thanks for the follow up. I ended up on this thread because I am experiencing what jpaulomotta mentioned, hangs while testing in development (never in production), where some times I see mentions of "Attempt to unlock a mutex which is locked by another thread" when I kill the local server. This all started happening after I started using puma/rack-timeout. I can't conclusively where the problem might be (was hoping it would be magically fixed by all that was being discussed here), so I wouldn't be able to give much detail as I don't know what to look for after the hang. I'm likely going way off topic of this thread now, am happy to reach out anyone directly who could direct me on how to investigate.

**Comment by kch** [comment]
@yaneeka and you tried using the 0.3.0.pre.beta.2 release?
